File: java/src/main/java/com/android/widget/PhotoViewer/helper/CustomGestureDetector.java
package com.android.widget.PhotoViewer.helper;

import android.content.Context;
import android.view.MotionEvent;
import android.view.ScaleGestureDetector;
import android.view.VelocityTracker;
import android.view.ViewConfiguration;

import com.android.widget.PhotoViewer.listener.OnGestureListener;

/**
 * Created by xuzhb on 2021/1/8
 * Desc:自定义手势检测
 */
public class CustomGestureDetector {

    private static final int INVALID_POINTER_ID = -1;

    private int mActivePointerId = INVALID_POINTER_ID;
    private int mActivePointerIndex = 0;
    private boolean mIsDragging;
    private float mLastTouchX;
    private float mLastTouchY;
    private VelocityTracker mVelocityTracker;      //速度追踪，追踪手指在滑动过程中的速度
    private final ScaleGestureDetector mDetector;  //缩放的手势检测
    private final OnGestureListener mListener;     //手势检测
    private final float mTouchSlop;        //系统所能识别的被认为是滑动的最小距离
    private final float mMinimumVelocity;  //执行一个fling的最小速度，以每秒像素数为单位

    public CustomGestureDetector(Context context, OnGestureListener listener) {
        final ViewConfiguration configuration = ViewConfiguration.get(context);
        mMinimumVelocity = configuration.getScaledMinimumFlingVelocity();
        mTouchSlop = configuration.getScaledTouchSlop();
        mListener = listener;
        ScaleGestureDetector.OnScaleGestureListener mScaleListener = new ScaleGestureDetector.OnScaleGestureListener() {

            @Override
            public boolean onScale(ScaleGestureDetector detector) {
                float scaleFactor = detector.getScaleFactor();
                if (Float.isNaN(scaleFactor) || Float.isInfinite(scaleFactor)) {
                    return false;
                }
                if (scaleFactor >= 0) {
                    mListener.onScale(scaleFactor, detector.getFocusX(), detector.getFocusY());
                }
                return true;
            }

            //缩放开始，返回结果表示是否处理后继的缩放事件，false表示不会执行onScale()
            @Override
            public boolean onScaleBegin(ScaleGestureDetector detector) {
                return true;
            }

            //缩放结束
            @Override
            public void onScaleEnd(ScaleGestureDetector detector) {
            }
        };
        mDetector = new ScaleGestureDetector(context, mScaleListener);
    }

    //是否正在缩放
    public boolean isScaling() {
        return mDetector.isInProgress();
    }

    //是否正在拖动
    public boolean isDragging() {
        return mIsDragging;
    }

    public boolean onTouchEvent(MotionEvent ev) {
        try {
            mDetector.onTouchEvent(ev);
            return processTouchEvent(ev);
        } catch (IllegalArgumentException e) {
            // Fix for support lib bug, happening when onDestroy is called
            return true;
        }
    }

    private boolean processTouchEvent(MotionEvent ev) {
        final int action = ev.getAction();
        switch (action & MotionEvent.ACTION_MASK) {
            case MotionEvent.ACTION_DOWN:
                mActivePointerId = ev.getPointerId(0);
                mVelocityTracker = VelocityTracker.obtain();
                if (mVelocityTracker != null) {
                    mVelocityTracker.addMovement(ev);
                }
                mLastTouchX = getActiveX(ev);
                mLastTouchY = getActiveY(ev);
                mIsDragging = false;
                break;
            case MotionEvent.ACTION_MOVE:
                final float x = getActiveX(ev);
                final float y = getActiveY(ev);
                final float dx = x - mLastTouchX;
                final float dy = y - mLastTouchY;
                if (!mIsDragging) {
                    mIsDragging = Math.sqrt((dx * dx) + (dy * dy)) >= mTouchSlop;  //判断是否正在滑动
                }
                if (mIsDragging) {
                    mListener.onDrag(dx, dy);
                    mLastTouchX = x;
                    mLastTouchY = y;
                    if (mVelocityTracker != null) {
                        mVelocityTracker.addMovement(ev);
                    }
                }
                break;
            case MotionEvent.ACTION_CANCEL:
                mActivePointerId = INVALID_POINTER_ID;
                //释放VelocityTracker
                if (mVelocityTracker != null) {
                    mVelocityTracker.recycle();
                    mVelocityTracker = null;
                }
                break;
            case MotionEvent.ACTION_UP:
                mActivePointerId = INVALID_POINTER_ID;
                if (mIsDragging) {
                    if (mVelocityTracker != null) {
                        mLastTouchX = getActiveX(ev);
                        mLastTouchY = getActiveY(ev);
                        mVelocityTracker.addMovement(ev);
                        mVelocityTracker.computeCurrentVelocity(1000);
                        //计算水平方向和竖直方向上手指滑动的速度
                        final float vX = mVelocityTracker.getXVelocity(), vY = mVelocityTracker.getYVelocity();
                        if (Math.max(Math.abs(vX), Math.abs(vY)) >= mMinimumVelocity) {
                            //如果超过了快速滑动的最小距离，则定性为一次快速滑动
                            mListener.onFling(mLastTouchX, mLastTouchY, -vX, -vY);
                        }
                    }
                }
                //释放VelocityTracker
                if (mVelocityTracker != null) {
                    mVelocityTracker.recycle();
                    mVelocityTracker = null;
                }
                break;
            //当屏幕上有多个点被按住，松开其中一个点时触发（即非最后一个点被放开时）触发
            case MotionEvent.ACTION_POINTER_UP:
                final int pointerIndex = getPointerIndex(ev.getAction());
                final int pointerId = ev.getPointerId(pointerIndex);
                if (pointerId == mActivePointerId) {
                    //重新设置一个新的位置
                    final int newPointerIndex = pointerIndex == 0 ? 1 : 0;
                    mActivePointerId = ev.getPointerId(newPointerIndex);
                    mLastTouchX = ev.getX(newPointerIndex);
                    mLastTouchY = ev.getY(newPointerIndex);
                }
                break;
        }
        mActivePointerIndex = ev.findPointerIndex(mActivePointerId != INVALID_POINTER_ID ? mActivePointerId : 0);
        return true;
    }

    private float getActiveX(MotionEvent ev) {
        try {
            return ev.getX(mActivePointerIndex);
        } catch (Exception e) {
            return ev.getX();
        }
    }

    private float getActiveY(MotionEvent ev) {
        try {
            return ev.getY(mActivePointerIndex);
        } catch (Exception e) {
            return ev.getY();
        }
    }

    private int getPointerIndex(int action) {
        return (action & MotionEvent.ACTION_POINTER_INDEX_MASK) >> MotionEvent.ACTION_POINTER_INDEX_SHIFT;
    }

}

File: kotlin/src/main/java/com/android/widget/PhotoViewer/helper/CustomGestureDetector.kt
package com.android.widget.PhotoViewer.helper

import android.content.Context
import android.view.MotionEvent
import android.view.ScaleGestureDetector
import android.view.VelocityTracker
import android.view.ViewConfiguration
import com.android.widget.PhotoViewer.listener.OnGestureListener
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.sqrt

/**
 * Created by xuzhb on 2021/3/30
 * Desc:自定义手势检测
 */
class CustomGestureDetector constructor(
    val mContext: Context,
    val mListener: OnGestureListener  //手势检测
) {

    companion object {
        private const val INVALID_POINTER_ID = -1
    }

    private var mActivePointerId: Int = INVALID_POINTER_ID
    private var mActivePointerIndex: Int = 0
    private var mIsDragging: Boolean = false
    private var mLastTouchX: Float = 0f
    private var mLastTouchY: Float = 0f
    private var mVelocityTracker: VelocityTracker? = null  //速度追踪，追踪手指在滑动过程中的速度
    private var mDetector: ScaleGestureDetector  //缩放的手势检测
    private var mTouchSlop: Int = 0        //系统所能识别的被认为是滑动的最小距离
    private var mMinimumVelocity: Int = 0  //执行一个fling的最小速度，以每秒像素数为单位

    init {
        val configuration = ViewConfiguration.get(mContext)
        mMinimumVelocity = configuration.scaledMinimumFlingVelocity
        mTouchSlop = configuration.scaledTouchSlop
        val scaleListener = object : ScaleGestureDetector.OnScaleGestureListener {

            //缩放开始，返回结果表示是否处理后继的缩放事件，false表示不会执行onScale()
            override fun onScaleBegin(detector: ScaleGestureDetector?): Boolean {
                return true
            }

            //缩放结束
            override fun onScaleEnd(detector: ScaleGestureDetector?) {
            }

            override fun onScale(detector: ScaleGestureDetector?): Boolean {
                detector?.let {
                    val scaleFactor = it.scaleFactor
                    if (java.lang.Float.isNaN(scaleFactor) || java.lang.Float.isInfinite(scaleFactor)) {
                        return false
                    }
                    if (scaleFactor >= 0) {
                        mListener.onScale(scaleFactor, it.focusX, it.focusY)
                    }
                }
                return true
            }
        }
        mDetector = ScaleGestureDetector(mContext, scaleListener)
    }

    //是否正在缩放
    fun isScaling(): Boolean = mDetector.isInProgress

    //是否正在拖动
    fun isDragging(): Boolean = mIsDragging

    fun onTouchEvent(ev: MotionEvent): Boolean {
        return try {
            mDetector.onTouchEvent(ev)
            processTouchEvent(ev)
        } catch (e: Exception) {
            // Fix for support lib bug, happening when onDestroy is called
            true
        }
    }

    private fun processTouchEvent(ev: MotionEvent): Boolean {
        val action = ev.action
        when (action and MotionEvent.ACTION_MASK) {
            MotionEvent.ACTION_DOWN -> {
                mActivePointerId = ev.getPointerId(0)
                mVelocityTracker = VelocityTracker.obtain()
                mVelocityTracker?.addMovement(ev)
                mLastTouchX = getActiveX(ev)
                mLastTouchY = getActiveY(ev)
                mIsDragging = false
            }
            MotionEvent.ACTION_MOVE -> {
                val x = getActiveX(ev)
                val y = getActiveY(ev)
                val dx = x - mLastTouchX
                val dy = y - mLastTouchY
                if (!mIsDragging) {
                    mIsDragging = sqrt((dx * dx) + (dy * dy)) >= mTouchSlop  //判断是否正在滑动
                }
                if (mIsDragging) {
                    mListener.onDrag(dx, dy)
                    mLastTouchX = x
                    mLastTouchY = y
                    mVelocityTracker?.addMovement(ev)
                }
            }
            MotionEvent.ACTION_CANCEL -> {
                mActivePointerId = INVALID_POINTER_ID
                //释放VelocityTracker
                mVelocityTracker?.recycle()
                mVelocityTracker = null
            }
            MotionEvent.ACTION_UP -> {
                mActivePointerId = INVALID_POINTER_ID
                if (mIsDragging) {
                    mVelocityTracker?.let {
                        mLastTouchX = getActiveX(ev)
                        mLastTouchY = getActiveY(ev)
                        it.addMovement(ev)
                        it.computeCurrentVelocity(1000)
                        //计算水平方向和竖直方向上手指滑动的速度
                        val vX = it.xVelocity
                        val vY = it.yVelocity
                        if (max(abs(vX), abs(vY)) >= mMinimumVelocity) {
                            //如果超过了快速滑动的最小距离，则定性为一次快速滑动
                            mListener.onFling(mLastTouchX, mLastTouchY, -vX, -vY)
                        }
                    }
                }
                //释放VelocityTracker
                mVelocityTracker?.recycle()
                mVelocityTracker = null
            }
            //当屏幕上有多个点被按住，松开其中一个点时触发（即非最后一个点被放开时）触发
            MotionEvent.ACTION_POINTER_UP -> {
                val pointerIndex = getPointerIndex(ev.action)
                val pointerId = ev.getPointerId(pointerIndex)
                if (pointerId == mActivePointerId) {
                    //重新设置一个新的位置
                    val newPointerIndex = if (pointerIndex == 0) 1 else 0
                    mActivePointerId = ev.getPointerId(newPointerIndex)
                    mLastTouchX = ev.getX(newPointerIndex)
                    mLastTouchY = ev.getX(newPointerIndex)
                }
            }
        }
        mActivePointerIndex = ev.findPointerIndex(if (mActivePointerId != INVALID_POINTER_ID) mActivePointerId else 0)
        return true
    }

    private fun getActiveX(ev: MotionEvent): Float {
        return try {
            ev.getX(mActivePointerIndex)
        } catch (e: Exception) {
            ev.x
        }
    }

    private fun getActiveY(ev: MotionEvent): Float {
        return try {
            ev.getY(mActivePointerIndex)
        } catch (e: Exception) {
            ev.y
        }
    }

    private fun getPointerIndex(action: Int): Int {
        return action and MotionEvent.ACTION_POINTER_INDEX_MASK shr MotionEvent.ACTION_POINTER_INDEX_SHIFT
    }

}